### DISCUSSION TITLE
Mapping Structured Mesh material labels to an Unstructured Mesh

### DISCUSSION THREAD

**Discussion by rafaelmarch3** [mfem/mfem#2532] [Q&A]
Hello, 

I'm having some issues mapping material labels from an structured array (3D image) into a MFEM Mesh object. Here's the way I'm doing it:

1) I have a structured cartesian mesh composed of nx x ny x nz voxels (It's a segmented image). 
2) I read the material labels (integer numbers), one for each voxel into an array of size nx*ny*nz. Let's say I have just two materials, labeled 1 and 2. 
3) Then, I instantiate MFEM's mesh with HEXAEDRON elements. 
4) I set the attributes by iterating over the number of elements of the MFEM mesh object and setting attribute of id "i" as the corresponding material from my material array. 

When I run the simulation (deformation of a heterogeneous volume) and open the vtk, the material array shows a completely different spatial distribution of my 1's and 2's. I believe this is related to the fact that MFEM is built for a generic unstructured mesh, so the global ids of each element of the mesh do not necessarily follow the index logic of my image file (first x, then y and then z). 

Any ideas on how to overcome this? One idea would be to look at the centroids of each cell as a way to map between the voxel values and the mesh elements. But I wonder if there would be any more straightforward way? 

Thank you, 
Rafael March.

**Comment by vladotomov** [comment]
Hi Rafael,

That's correct, it's safer to not assume particular ordering of the mesh elements. I would create a `Coefficient` that computes the physical coordinates of the used point, and extract the corresponding label. Then this `Coefficient` can be projected on a piecewise constant `GridFunction`.

Vladimir

**Comment by rcarson3** [comment]
@rafaelmarch3 so this shouldn't be an issue. I do something similar to this within my crystal plasticity code where I need to assign a grain ID to each element if I'm using the auto mesh generator within MFEM. Here's an example of how I do it in one of my mesh generation programs: https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/scripts/meshing/mesh_generator.cpp or here in my main driver https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/mechanics_driver.cpp#L234-L269. The key is to make sure in your case the material ID / voxel label is read into the serial mesh with the x direction iterating the fastest, then the y direction and finally the z direction. Once your serial mesh has the element IDs assigned then you can create the parallel mesh and the IDs will be correctly assigned.

Also, the voxel thing shouldn't be an issue if your mesh has the same number of elements in each x,y, and z direction as your voxel image. I do this all the time in my simulations and the mapping works out just fine. If that's not the case then you would have to work on the mapping aspect of things. One thing you might want to do is double check and make sure your material array actually has the correct ordering of things that you expect. I've run into this issue before where it didn't correspond to the x->y->z type ordering so my element attributes were completely off when I read them into my simulations.
